src/ripple/basics/impl/Number.cpp

```diff
@@ -0,0 +1,732 @@
+
+Number::rounding_mode
+Number::getround()
+{
+    return mode_;
+}
+
+Number::rounding_mode
+Number::setround(rounding_mode mode)
+{
+    return std::exchange(mode_, mode);
+}
+
+// Guard
+
+// The Guard class is used to tempoarily add extra digits of
+// preicision to an operation.  This enables the final result
+// to be correctly rounded to the internal precision of Number.
+
+class Number::Guard
+{
+    std::uint64_t digits_;   // 16 decimal guard digits
+    std::uint8_t xbit_ : 1;  // has a non-zero digit been shifted off the end
+    std::uint8_t sbit_ : 1;  // the sign of the guard digits
+
+public:
+    explicit Guard() : digits_{0}, xbit_{0}, sbit_{0}
+    {
+    }
+
+    // set & test the sign bit
+    void
+    set_positive() noexcept;
+    void
+    set_negative() noexcept;
+    bool
+    is_negative() const noexcept;
+
+    // add a digit
+    void
+    push(unsigned d) noexcept;
+
+    // recover a digit
+    unsigned
+    pop() noexcept;
+
+    // Indicate round direction:  1 is up, -1 is down, 0 is even
+    // This enables the client to round towards nearest, and on
+    // tie, round towards even.
+    int
+    round() noexcept;
+};
+
+inline void
+Number::Guard::set_positive() noexcept
+{
+    sbit_ = 0;
+}
+
+inline void
+Number::Guard::set_negative() noexcept
+{
+    sbit_ = 1;
+}
+
+inline bool
+Number::Guard::is_negative() const noexcept
+{
+    return sbit_ == 1;
+}
+
+inline void
+Number::Guard::push(unsigned d) noexcept
+{
+    xbit_ = xbit_ || (digits_ & 0x0000'0000'0000'000F) != 0;
+    digits_ >>= 4;
+    digits_ |= (d & 0x0000'0000'0000'000FULL) << 60;
+}
+
+inline unsigned
+Number::Guard::pop() noexcept
+{
+    unsigned d = (digits_ & 0xF000'0000'0000'0000) >> 60;
+    digits_ <<= 4;
+    return d;
+}
+
+// Returns:
+//     -1 if Guard is less than half
+//      0 if Guard is exactly half
+//      1 if Guard is greater than half
+int
+Number::Guard::round() noexcept
+{
+    auto mode = Number::getround();
+
+    if (mode == towards_zero)
+        return -1;
+
+    if (mode == downward)
+    {
+        if (sbit_)
+        {
+            if (digits_ > 0 || xbit_)
+                return 1;
+        }
+        return -1;
+    }
+
+    if (mode == upward)
+    {
+        if (sbit_)
+            return -1;
+        if (digits_ > 0 || xbit_)
+            return 1;
+        return -1;
+    }
+
+    // assume round to nearest if mode is not one of the predefined values
+    if (digits_ > 0x5000'0000'0000'0000)
+        return 1;
+    if (digits_ < 0x5000'0000'0000'0000)
+        return -1;
+    if (xbit_)
+        return 1;
+    return 0;
+}
+
+// Number
+
+constexpr Number one{1000000000000000, -15, Number::unchecked{}};
+
+void
+Number::normalize()
+{
+    if (mantissa_ == 0)
+    {
+        *this = Number{};
+        return;
+    }
+    bool const negative = (mantissa_ < 0);
+    auto m = static_cast<std::make_unsigned_t<rep>>(mantissa_);
+    if (negative)
+        m = -m;
```

Comment:
In several algorithms, we negate a number. Negating the largest negative value doesn't actually negate it. Can you add a comment near each negation explaining why we never see the largest negative value or what the consequences of that value would be for the algorithm?